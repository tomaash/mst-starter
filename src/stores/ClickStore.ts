import { types } from 'mobx-state-tree'

export const ClickStore = types
  .model('ClickStore', {
    count: 0
  })
  .actions(self => {
    function inc() {
      self.count += 1
    }
    return {
      inc
    }
  })

export type IClickStore = typeof ClickStore.Type
