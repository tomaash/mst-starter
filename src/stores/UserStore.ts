import { values } from 'mobx'
import { types, getParent, flow } from 'mobx-state-tree'
import axios from 'axios'

export const User = types.model('User', {
  id: types.identifierNumber,
  first_name: types.string,
  last_name: types.string,
  avatar: types.string,
  email: types.string
})

export type IUser = typeof User.Type

export const UserStore = types
  .model('UserStore', {
    isLoading: true,
    users: types.array(User)
  })
  .views(self => ({
    get sortedUsers() {
      return sortUsers(values(self.users))
    }
  }))
  .actions(self => {
    function markLoading(loading) {
      self.isLoading = loading
    }
    function updateUsers(json) {
      self.users = json.data.data
      // json.data.data.forEach(user => {
      //   self.users.put(user)
      // })
    }
    const loadUsers = flow(function* loadUsers() {
      try {
        const json = yield axios.get('https://reqres.in/api/users')
        updateUsers(json)
        markLoading(false)
      } catch (err) {
        console.error('Failed to load books ', err)
      }
    })
    return {
      updateUsers,
      loadUsers
    }
  })

export type IUserStore = typeof UserStore.Type

function sortUsers(books) {
  return books.sort((a, b) =>
    a.last_name > b.last_name ? 1 : a.last_name === b.last_name ? 0 : -1
  )
}
