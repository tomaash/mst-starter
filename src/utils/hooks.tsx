import { useEffect } from 'react'
export const useMount = cb => {
  return useEffect(cb, [])
}
