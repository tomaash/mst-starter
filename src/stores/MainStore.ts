import React from 'react'
import { types } from 'mobx-state-tree'
import { UserStore } from './UserStore'
import { MobXProviderContext } from 'mobx-react'
import { ClickStore } from './ClickStore'

export const MainStore = types.model('MainStore', {
  userStore: types.optional(UserStore, {
    isLoading: true,
    users: []
  }),
  clickStore: types.optional(ClickStore, {
    count: 0
  })
})

export type IMainStore = typeof MainStore.Type

export function useStore() {
  return React.useContext(MobXProviderContext) as { mainStore: IMainStore }
}
