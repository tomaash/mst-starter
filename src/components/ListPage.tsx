import React from 'react'
import { observer, useLocalStore } from 'mobx-react-lite'
import { useDidMount } from 'rooks'
import {
  Table,
  TableBody,
  TableCell,
  TableHead,
  TableRow,
  Button
} from '@material-ui/core'
import { RouteComponentProps } from '@reach/router'
import { useStore } from '~/stores/MainStore'

export const ListPageComponent = observer<RouteComponentProps>(props => {
  const { mainStore } = useStore()
  const { userStore, clickStore } = mainStore

  useDidMount(() => {
    userStore.loadUsers()
  })
  function handleClick() {
    clickStore.inc()
  }

  const localState = useLocalStore(() => ({
    count: 0,
    inc() {
      localState.count += 1
    }
  }))
  return (
    <>
      <Table>
        <TableHead>
          <TableRow>
            <TableCell>First Name</TableCell>
            <TableCell>Last Name</TableCell>
          </TableRow>
        </TableHead>
        <TableBody>
          {userStore.users.map(item => {
            return (
              <TableRow data-testid='table-row' key={item.id}>
                <TableCell>{item.first_name}</TableCell>
                <TableCell>{item.last_name}</TableCell>
              </TableRow>
            )
          })}
        </TableBody>
      </Table>
      <Button onClick={handleClick}>
        Click store count: {clickStore.count}
      </Button>
      <br />
      <Button onClick={localState.inc}>
        Click local count: {localState.count}
      </Button>
    </>
  )
})

// TODO: refactor to smart & dumb
export const ListPage = ListPageComponent
