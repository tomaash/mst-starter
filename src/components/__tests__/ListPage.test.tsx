import React from 'react'
import { renderText, squashHtmlTags, startPolly } from '~/utils/testUtils'
import { Provider } from 'mobx-react'
import { ListPage } from '../ListPage'
import { MainStore } from '../../stores/MainStore'

describe('<ListPage />', () => {
  it('renders component', () => {
    const componentSetup = (
      <Provider mainStore={MainStore.create()}>
        <ListPage />
      </Provider>
    )

    const text = renderText(componentSetup)
    expect(squashHtmlTags(text)).toMatchSnapshot()
  })

  it('renders component with data', async () => {
    const mainStore = MainStore.create({
      userStore: {
        isLoading: false,
        users: [
          {
            id: 1,
            first_name: 'George',
            last_name: 'Bluth',
            avatar:
              'https://s3.amazonaws.com/uifaces/faces/twitter/calebogden/128.jpg',
            email: 'george.bluth@reqres.in'
          },
          {
            id: 2,
            first_name: 'Janet',
            last_name: 'Weaver',
            avatar:
              'https://s3.amazonaws.com/uifaces/faces/twitter/josephstein/128.jpg',
            email: 'janet.weaver@reqres.in'
          },
          {
            id: 3,
            first_name: 'Emma',
            last_name: 'Wong',
            avatar:
              'https://s3.amazonaws.com/uifaces/faces/twitter/olegpogodaev/128.jpg',
            email: 'emma.wong@reqres.in'
          }
        ]
      },
      clickStore: { count: 3 }
    })
    const componentSetup = (
      <Provider mainStore={mainStore}>
        <ListPage />
      </Provider>
    )
    expect(renderText(componentSetup)).toMatchSnapshot()
  })
})
