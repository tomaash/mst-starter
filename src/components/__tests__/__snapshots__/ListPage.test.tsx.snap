// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ListPage /> renders component 1`] = `
"
First Name
Last Name
Click store count: 0
Click local count: 0
"
`;

exports[`<ListPage /> renders component with data 1`] = `
"
First Name
Last Name
George
Bluth
Janet
Weaver
Emma
Wong
Click store count: 3
Click local count: 0
"
`;
